File: algortihms-java/src/com/algo/tree/traversal/InOrderTraversal.java
package com.algo.tree.traversal;


import com.algo.tree.Node;

import java.util.Stack;

public class InOrderTraversal {
    public static void main(String[] args) {
        Node node1 = new Node(1);
        Node node2 = new Node(2);
        Node node3 = new Node(3);
        Node node4 = new Node(4);
        Node node5 = new Node(5);
        Node node6 = new Node(6);
        node1.setLeft(node2);
        node1.setRight(node3);
        node2.setLeft(node4);
        node2.setRight(node5);
        node3.setLeft(node6);
//        printInOrder1(node1);
        printInOrder2(node1);
    }

    private static void printInOrder1(Node root) {
        if (root == null) {
            return;
        }
        printInOrder1(root.getLeft());
        System.out.print(root.getData() + "\t");
        printInOrder1(root.getRight());
    }

    private static void printInOrder2(Node root) {
        if (root == null) {
            throw new IllegalArgumentException("Root node cannot be null");
        }
        Stack<Node> stack = new Stack<>();
        while (true) {
            while (root != null) {
                stack.push(root);
                root = root.getLeft();
            }
            if (!stack.isEmpty()) {
                Node current = stack.pop();
                System.out.print(current.getData() + "\t");
                root = current.getRight();
            } else {
                break;
            }
        }
    }


}

File: algorithms-python/trees/traverse/inorder_traversal.py
from trees.node import Node
from collections import deque

'''
                        1
                    /      \
                    2        3
                  /   \        \
                4        5       6
                                /
                               7           
'''


def inorder2(root: Node) -> None:
    if root is None:
        return
    stack: deque = deque()
    while True:
        while root is not None:
            stack.append(root)
            root = root.left
        if stack:
            node: Node = stack.pop()
            print(node.data, end="\t")
            root = node.right
        else:
            break


def inorder1(root: Node) -> None:
    if root is None:
        return
    inorder1(root.left)
    print(root.data, end="\t")
    inorder1(root.right)


def main():
    node1: Node = Node(1)
    node2: Node = Node(2)
    node3: Node = Node(3)
    node4: Node = Node(4)
    node5: Node = Node(5)
    node6: Node = Node(6)
    node7: Node = Node(7)
    node1.left = node2
    node1.right = node3
    node2.left = node4
    node2.right = node5
    node3.right = node6
    node6.left = node7
    inorder1(node1)
    print()
    inorder2(node1)
    print()


if __name__ == '__main__':
    main()
